keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,215 @@
+
+Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.
+
+As for Go types, we only accept one union per structure, which means there can
+be only one discriminator, and all other `oneof` fields are assumed to belong to
+the same union.
+
+### Discriminator
+
+For backward compatibility reasons, discriminators can be added to existing
+union structures as an optional string. This has a nice property that it's going
+to allow conflict detection when the selected union field is changed.
+
+The value of the discriminator is going to be set automatically by the apiserver
+when a new field is changed in the union. It will be set to the value of the
+`x-kubernetes-union-fields-discriminated` extension for that specific field.
+
+When the value of the discriminator is explicitly changed by the client, it
+will be interpreted as an intention to clear all the other fields. See section
+below.
+
+If the object had exactly one of the union field set, and that field is removed,
+the discriminator will be cleared.
+
+### Normalizing on updates
+
+A "normalization process" will run automatically for all creations and
+modifications (either with put or patch). It will happen automatically in order
+to clear fields and update discriminators. This process will run for both
+core-types and CRDs. It will take place before validation. The sent object
+doesn't have to be valid, but fields will be cleared in order to make it valid.
+This process will also happen before fields tracking (server-side apply), so
+changes in discriminator, even if implicit, will be owned by the client making
+the update (and may result in conflicts).
+
+This process works as follows:
+- If there is a discriminator, and its value has changed, clear all fields but
+  the one specified by the discriminator,
+- If there is no discriminator, or if its value hasn't changed,
+  - if there is exactly one field set, set the discriminator (if exists) to that
+    value. Otherwise,
+  - compare the fields set before and after. If there is exactly one field
+    added, set the discriminator (if present) to that value, and remove all
+    other fields. if more than one field has been added, leave the process so
+    that validation will fail.
+
+#### "At most one" versus "exactly one"
+
+The goal of this proposal is not to change the validation, but to help client
+clear other fields in the union. Validation should be implemented for in-tree
+types as it is today, or through "oneOf" properties in CRDs.
+
+In other word, this is proposing to implement "at most one", and the exactly one
+should be provided through another layer of validation (separated).
+
+#### Clearing all the fields
+
+Since the system is trying to do the right thing, it can be hard to "clear
+everything". In that case, each API could decide to have their own "Nothing"
+value in the discriminator, which will automatically trigger a clearing of all
+fields beside "Nothing".
```

Comment:
An explicit empty string would not do?

Reply:
We're assuming that clients that deserialize in golang and are not aware of a newly introduced discriminator would reset that string by accident. We don't want to infer intent from that.

Reply:
makes sense.

Reply:
If the "Nothing" value is customizable, how will that happen via tags?

Reply:
`// +discriminator:nothingValue="Nothing"` or some such, presumably.  We've got at least one union in kube today that requires this.

It'd be nice to be able to explicitly say that you don't want this behavior though, for new APIs going forward.

Reply:
We already have one of these and it works very well:  `DeploymentStrategy`.

`Recreate` is the "nothing", `RolloutStrategy` is the one that has an associated value. We should let it up to validation to verify which values are accepted for the discriminator, but I don't think we should build anything more complicated here.